File: 3B/cs343/A5/philosopher.cc
#include "philosopher.h"
#include "table.h"
#include "printer.h"

Philosopher::Philosopher(unsigned int id, unsigned int noodles,
                         Table& table, Printer& prt)
  : id(id), noodles(noodles), table(table), prt(prt) {}

void Philosopher::main() {
  while (true) {
    // Start Hungry
    prt.print(id, HUNGRY);
    yield(rand() % 5);

    // Acquire forks
    table.pickup(id);

    // Eat
    unsigned int bite = (rand() % 5) + 1;
    if (bite > noodles) bite = noodles;
    noodles -= bite;
    prt.print(id, EATING, bite, noodles);
    yield(rand() % 5);

    // Put down forks
    table.putdown(id);

    // Check if done noodles
    if (noodles <= 0) break;

    // Think
    prt.print(id, THINKING);
    yield(rand() % 20);
  }

  // Terminate
  prt.print(id, FINISHED);
}

File: 3B/cs343/A5/Philosopher.java

public class Philosopher extends Thread {
	public enum States {THINKING, HUNGRY, EATING, WAITING, FINISHED};

	private int id;
	private int noodles;
	private Table table;
	private Printer prt;

	@Override
	public void run() {
		while (true) {
			// Start Hungry
			prt.print(id, States.HUNGRY);
			int time = (int)(Math.random() * 5);
			for (int i = 0; i < time; i++) {
				yield();
			}

			// Acquire Forks
			table.pickup(id);

			// Eat
			int bite = (int)(Math.random() * 5) + 1;
			if (bite > noodles) {
				bite = noodles;
			}
			noodles -= bite;
			prt.print(id, States.EATING, bite, noodles);
			time = (int)(Math.random() * 5);
			for (int i = 0; i < time; i++) {
				yield();
			}

			// Put down forks
			table.putdown(id);

			// Check if done noodles
			if (noodles <= 0) {
				break;
			}

			// Think
			time = (int)(Math.random() * 20);
			for (int i = 0; i < time; i++) {
				yield();
			}
		}
		
		// Terminate
		prt.print(id, States.FINISHED);
	}

	public Philosopher(int id, int noodles, Table table, Printer prt) {
		this.id = id;
		this.noodles = noodles;
		this.table = table;
		this.prt = prt;
	}
}
